every package we provide is documented

boyska opened the pull request:
This is created using 
```bash
sphinx-apidoc -EM -H libreant . -o doc/source/api -d1 webant/ utils/ *.py cli/ */test/
```
plus some cleaning (removing doctests from generated files).

While it creates some unwanted sections (like empty _Subpackages_ and _Submodules_) the output is quite good, and I like how "reproducible it is" using `sphinx-apidoc`, which makes adding new packages to documentation easy.

To test it, you must run `python setup.py build_sphinx` and then `firefox build/sphinx/html/index.html`

What you should find is a much extended `API` section, which now contains every package we ship. Of course, not everything is properly documented, and the fact that the API generation was missing for so much time means that some docstring is not formatted as expected. I won't fix those bugs in this pull request.

ael-code commented:
I think this is documenting too much:
cli, msgfmt, setup should not be documented cause they do not provide docstring at all.

I will not provide also docs for utils and webant.

A lot of auto package doc contains empty `Module contents` and `Submodule` sections.

boyska commented:
ael wrote:
>I think this is documenting too much:
>cli, msgfmt, setup should not be documented cause they do not provide
>docstring at all.
>I will not provide also docs for utils and webant.

good point. I'll remove them.
So the rationale is that we should document only packages that are meant
to be used by a _user_ of libreant, not a _developer_ of libreant; am I
correct?

>A lot of auto package doc contains empty `Module contents` and
>`Submodule` sections.

Yes, this is mentioned in the PR description. I favored
"reproducibility" to "accuracy", because writing API .rst files by hand
is so boring, I supposed that the only way to have an updated one is
making it a simple and reproducible process.

Of course having it reproducible and with more accurate content is a
plus, but I haven't figured out how to do it.

boyska commented:
I think that the new code (and the new way of generating it) addresses all concerns:
* "useless" packages are not documented at all.
* the generation is automatic
* the "submodules" part actually contains submodules, so it's not useless

boyska commented:
ooops!
on rtfd  https://libreant.readthedocs.org/en/latest/api/archivant.html#submodules
there is no sign of API generation. What's going on?

ael-code commented:
On 11/05/2015 11:26 AM, BoySka wrote:
> ooops!
> on rtfd
> https://libreant.readthedocs.org/en/latest/api/archivant.html#submodules
> there is no sign of API generation. What's going on?
> 

Yeap I noticed that.

ael-code commented:
On 11/05/2015 11:32 AM, ael wrote:
> On 11/05/2015 11:26 AM, BoySka wrote:
>> ooops!
>> on rtfd
>> https://libreant.readthedocs.org/en/latest/api/archivant.html#submodules
>> there is no sign of API generation. What's going on?
>>
> 
> Yeap I noticed that.
> 


This is the rtfd build log
```
 Running Sphinx v1.3.1
making output directory...
loading translations [en]... done
building [mo]: targets for 0 po files that are out of date
building [readthedocs]: targets for 13 source files that are out of date
updating environment: 13 added, 0 changed, 0 removed
reading sources... [  7%] about
reading sources... [ 15%] api/archivant
reading sources... [ 23%] api/conf
reading sources... [ 30%] api/libreantdb
reading sources... [ 38%] api/modules
reading sources... [ 46%] api/presets
reading sources... [ 53%] api/users
reading sources... [ 61%] architecture
reading sources... [ 69%] developer
reading sources... [ 76%] documentation
reading sources... [ 84%] index
reading sources... [ 92%] librarian
reading sources... [100%] sysadmin

/home/docs/checkouts/readthedocs.org/user_builds/libreant/checkouts/latest/doc/source/api/archivant.rst:4:
WARNING: autodoc: failed to import module u'archivant'; the following
exception was raised:
Traceback (most recent call last):
  File
"/home/docs/checkouts/readthedocs.org/user_builds/libreant/envs/latest/local/lib/python2.7/site-packages/sphinx/ext/autodoc.py",
line 385, in import_object
    __import__(self.modname)
  File
"/home/docs/checkouts/readthedocs.org/user_builds/libreant/checkouts/latest/archivant/__init__.py",
line 1, in <module>
    from archivant import Archivant
  File
"/home/docs/checkouts/readthedocs.org/user_builds/libreant/checkouts/latest/archivant/archivant.py",
line 3, in <module>
    from elasticsearch import Elasticsearch
ImportError: No module named elasticsearch
/home/docs/checkouts/readthedocs.org/user_builds/libreant/checkouts/latest/doc/source/api/archivant.rst:12:
WARNING: autodoc: failed to import module u'archivant.archivant'; the
following exception was raised:
Traceback (most recent call last):
  File
"/home/docs/checkouts/readthedocs.org/user_builds/libreant/envs/latest/local/lib/python2.7/site-packages/sphinx/ext/autodoc.py",
line 385, in import_object
    __import__(self.modname)
  File
"/home/docs/checkouts/readthedocs.org/user_builds/libreant/checkouts/latest/archivant/__init__.py",
line 1, in <module>
    from archivant import Archivant
  File
"/home/docs/checkouts/readthedocs.org/user_builds/libreant/checkouts/latest/archivant/archivant.py",
line 3, in <module>
    from elasticsearch import Elasticsearch
ImportError: No module named elasticsearch
/home/docs/checkouts/readthedocs.org/user_builds/libreant/checkouts/latest/doc/source/api/archivant.rst:17:
WARNING: autodoc: failed to import module u'archivant.exceptions'; the
following exception was raised:
Traceback (most recent call last):
  File
"/home/docs/checkouts/readthedocs.org/user_builds/libreant/envs/latest/local/lib/python2.7/site-packages/sphinx/ext/autodoc.py",
line 385, in import_object
    __import__(self.modname)
  File
"/home/docs/checkouts/readthedocs.org/user_builds/libreant/checkouts/latest/archivant/__init__.py",
line 1, in <module>
    from archivant import Archivant
  File
"/home/docs/checkouts/readthedocs.org/user_builds/libreant/checkouts/latest/archivant/archivant.py",
line 3, in <module>
    from elasticsearch import Elasticsearch
ImportError: No module named elasticsearch
/home/docs/checkouts/readthedocs.org/user_builds/libreant/checkouts/latest/conf/config_utils.py:docstring
of conf.config_utils.load_configs:5: WARNING: Inline interpreted text or
phrase reference start-string without end-string.
/home/docs/checkouts/readthedocs.org/user_builds/libreant/checkouts/latest/doc/source/api/conf.rst:17:
WARNING: autodoc: failed to import module u'conf.defaults'; the
following exception was raised:
Traceback (most recent call last):
  File
"/home/docs/checkouts/readthedocs.org/user_builds/libreant/envs/latest/local/lib/python2.7/site-packages/sphinx/ext/autodoc.py",
line 385, in import_object
    __import__(self.modname)
  File
"/home/docs/checkouts/readthedocs.org/user_builds/libreant/checkouts/latest/conf/defaults.py",
line 1, in <module>
    from passlib.hash import pbkdf2_sha256
ImportError: No module named passlib.hash
/home/docs/checkouts/readthedocs.org/user_builds/libreant/checkouts/latest/doc/source/api/libreantdb.rst:4:
WARNING: autodoc: failed to import module u'libreantdb'; the following
exception was raised:
Traceback (most recent call last):
  File
"/home/docs/checkouts/readthedocs.org/user_builds/libreant/envs/latest/local/lib/python2.7/site-packages/sphinx/ext/autodoc.py",
line 385, in import_object
    __import__(self.modname)
  File
"/home/docs/checkouts/readthedocs.org/user_builds/libreant/checkouts/latest/libreantdb/__init__.py",
line 1, in <module>
    from api import DB
  File
"/home/docs/checkouts/readthedocs.org/user_builds/libreant/checkouts/latest/libreantdb/api.py",
line 1, in <module>
    from elasticsearch import NotFoundError
ImportError: No module named elasticsearch
/home/docs/checkouts/readthedocs.org/user_builds/libreant/checkouts/latest/doc/source/api/libreantdb.rst:12:
WARNING: autodoc: failed to import module u'libreantdb.api'; the
following exception was raised:
Traceback (most recent call last):
  File
"/home/docs/checkouts/readthedocs.org/user_builds/libreant/envs/latest/local/lib/python2.7/site-packages/sphinx/ext/autodoc.py",
line 385, in import_object
    __import__(self.modname)
  File
"/home/docs/checkouts/readthedocs.org/user_builds/libreant/checkouts/latest/libreantdb/__init__.py",
line 1, in <module>
    from api import DB
  File
"/home/docs/checkouts/readthedocs.org/user_builds/libreant/checkouts/latest/libreantdb/api.py",
line 1, in <module>
    from elasticsearch import NotFoundError
ImportError: No module named elasticsearch
/home/docs/checkouts/readthedocs.org/user_builds/libreant/checkouts/latest/doc/source/api/presets.rst:4:
WARNING: __all__ should be a list of strings, not [<class
'presets.presetManager.PresetManager'>] (in module presets) -- ignoring
__all__
/home/docs/checkouts/readthedocs.org/user_builds/libreant/checkouts/latest/presets/presetManager.py:docstring
of presets.presetManager.Schema:17: WARNING: Block quote ends without a
blank line; unexpected unindent.
/home/docs/checkouts/readthedocs.org/user_builds/libreant/checkouts/latest/presets/presetManager.py:docstring
of presets.presetManager.Schema:19: ERROR: Unexpected indentation.
/home/docs/checkouts/readthedocs.org/user_builds/libreant/checkouts/latest/presets/presetManager.py:docstring
of presets.presetManager.Schema:22: WARNING: Block quote ends without a
blank line; unexpected unindent.
/home/docs/checkouts/readthedocs.org/user_builds/libreant/checkouts/latest/presets/presetManager.py:docstring
of presets.presetManager.Schema:23: WARNING: Block quote ends without a
blank line; unexpected unindent.
/home/docs/checkouts/readthedocs.org/user_builds/libreant/checkouts/latest/doc/source/api/users.rst:4:
WARNING: autodoc: failed to import module u'users'; the following
exception was raised:
Traceback (most recent call last):
  File
"/home/docs/checkouts/readthedocs.org/user_builds/libreant/envs/latest/local/lib/python2.7/site-packages/sphinx/ext/autodoc.py",
line 385, in import_object
    __import__(self.modname)
  File
"/home/docs/checkouts/readthedocs.org/user_builds/libreant/checkouts/latest/users/__init__.py",
line 1, in <module>
    from peewee import SqliteDatabase
ImportError: No module named peewee
/home/docs/checkouts/readthedocs.org/user_builds/libreant/checkouts/latest/doc/source/api/users.rst:12:
WARNING: autodoc: failed to import module u'users.api'; the following
exception was raised:
Traceback (most recent call last):
  File
"/home/docs/checkouts/readthedocs.org/user_builds/libreant/envs/latest/local/lib/python2.7/site-packages/sphinx/ext/autodoc.py",
line 385, in import_object
    __import__(self.modname)
  File
"/home/docs/checkouts/readthedocs.org/user_builds/libreant/checkouts/latest/users/__init__.py",
line 1, in <module>
    from peewee import SqliteDatabase
ImportError: No module named peewee
/home/docs/checkouts/readthedocs.org/user_builds/libreant/checkouts/latest/doc/source/api/users.rst:17:
WARNING: autodoc: failed to import module u'users.models'; the following
exception was raised:
Traceback (most recent call last):
  File
"/home/docs/checkouts/readthedocs.org/user_builds/libreant/envs/latest/local/lib/python2.7/site-packages/sphinx/ext/autodoc.py",
line 385, in import_object
    __import__(self.modname)
  File
"/home/docs/checkouts/readthedocs.org/user_builds/libreant/checkouts/latest/users/__init__.py",
line 1, in <module>
    from peewee import SqliteDatabase
ImportError: No module named peewee
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
writing output... [  7%] about
writing output... [ 15%] api/archivant
writing output... [ 23%] api/conf
writing output... [ 30%] api/libreantdb
writing output... [ 38%] api/modules
writing output... [ 46%] api/presets
writing output... [ 53%] api/users
writing output... [ 61%] architecture
writing output... [ 69%] developer
writing output... [ 76%] documentation
writing output... [ 84%] index
writing output... [ 92%] librarian
writing output... [100%] sysadmin

generating indices... genindex py-modindex
highlighting module code... [ 50%] presets.presetManager
highlighting module code... [100%] conf.config_utils

writing additional pages... search
copying static files... WARNING: html_static_path entry
'/home/docs/checkouts/readthedocs.org/readthedocs/templates/sphinx/_static'
does not exist
done
copying extra files... done
dumping search index in English (code: en) ... done
dumping object inventory... done
build succeeded, 16 warnings.
Copying readthedocs-dynamic-include.js_t... done
Copying readthedocs-data.js_t... done

```

boyska commented:
ok, I think the problem is clearer now. Trying to fix on rtfd's side

boyska commented:
ok, it works now (enabling the `do python setup.py install` thing on Admin->advanced settings was enough). enjoy.
